### PR TITLE
test/e2e: Reduce need to use local kubectl

### PR DESCRIPTION
**Pull request by okartau**
Use POST method instead of running local kunectl.
Use ExecCommandInContainerWithFullOutput() instead of RunKubectl().
PodExec() takes additional framework arg, passed down in call chain.
New variants of two functions in e2e/framework/pods.go:
ExecWithOptions, ExecCommandInContainerWithFullOutput
with added ClientSet, Namespace arguments.
VerifyExecInPodFail uses different error code cast as original
one causes test code Panic if used with new call method.

/kind feature

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Removed dependency on kubectl from several storage E2E tests
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by BenTheElder** [comment]
we have conformance tests specifically about using kubectl and that's unlikely to change.

if you want to not use a local kubectl, have you considered using the confromance image and / or sonobuoy?

**Comment by BenTheElder** [comment]
I see, I might have suggested one that does NOT import k8s.io/kubernetes ...

**Comment by pohly** [comment]
> I see, I might have suggested one that does NOT import k8s.io/kubernetes ...

Another test framework? But then it wouldn't have the same tests readily available, would it?

Anyway, work is underway to move the test/e2e/framework into staging, at which point importing it will be easier.

**Review by pohly — CHANGES_REQUESTED** [review]

**Comment by pohly** [comment]
/lgtm

I'll let others approve, to ensure that we have consensus that this is a useful change.

File: test/e2e/framework/volume/fixtures.go
Comment (pohly) [inline test/e2e/framework/volume/fixtures.go]: /lgtm cancel

I think this is another case of a redundant `client` parameter. Remove the parameter and use `f.ClientSet` instead in the call to `runVolumeTesterPod`.

Comment (okartau) [reply test/e2e/framework/volume/fixtures.go]: made that change and pushed.
Should we also delete cs arg from test/e2e/storage/flexvolume.go:testFlexVolume, which now becomes unused. By same logic, passing cs is redundant already in current code if we assume it is always set to f.ClientSet, and framework arg is there now already

Comment (pohly) [reply test/e2e/framework/volume/fixtures.go]: > Should we also delete cs arg from test/e2e/storage/flexvolume.go:testFlexVolume, which now becomes unused.

Yes.

> By same logic, passing cs is redundant already in current code if we assume it is always set to f.ClientSet, and framework arg is there now already

Correct.

**Review by pohly — APPROVED** [review]
/lgtm

**Comment by okartau** [comment]
My recent force-push from 5h ago was done after manual merge-to-HEAD conflict resolve.

File: test/e2e/framework/pods.go
Comment (cofyc) [inline test/e2e/framework/pods.go]: I have two questions:

1)  why we need two new methods on PodClient? These two methods do not access PodClient fields. And we already have [exec methods on Framework](https://github.com/kubernetes/kubernetes/blob/ed5b03831385b0071b127ff6c329ae344b9780f6/test/e2e/framework/exec_util.go#L50), I see the usage of `PodClient()` [here](https://github.com/kubernetes/kubernetes/pull/84042/files#diff-0d21a89b3adc375caf1b557b014e1111R68), It still use the framework ClientSet. 

2) If we need to, I think it's better to have some subroutines and reuse code.

Comment (pohly) [reply test/e2e/framework/pods.go]: I might have suggested that to @okartau as one option before he started working on the code. But as you noticed, in the end the resulting methods are not related to `PodClient`, which indeed indicates that they belong elsewhere.

But before we discuss where a better place for them might be, do we actually still need this PR? It was supposed to solve https://github.com/kubernetes/kubernetes/issues/83913, but as you pointed out in https://github.com/kubernetes/kubernetes/issues/83913#issuecomment-548683208, that problem was already solved (differently, but still solved) in https://github.com/kubernetes/kubernetes/pull/84444, correct?

In that case we can simply close this PR unless someone sees some value in these enhancements.

Comment (okartau) [reply test/e2e/framework/pods.go]: I am ok with closing.

Comment (cofyc) [reply test/e2e/framework/pods.go]: HostExec is an alternative way to `ssh.SSH`. To avoid using kubectl, `*Framework` is still needed to access ExecWithOptions/ExecCommandInContainerWithFullOutput methods.

Comment (pohly) [reply test/e2e/framework/pods.go]: @cofyc: so you are saying that https://github.com/kubernetes/kubernetes/pull/84444 does not remove the kubectl dependency completely and therefore this PR is still needed? That part is not clear to me.

But we have the source, so let's use that... :smile:

In current Kubernetes master, https://github.com/kubernetes/kubernetes/blob/08410cbf06f1c8649157407c660b1c5cb528947b/test/e2e/storage/testsuites/multivolume.go#L368 uses this chain of calls:
- https://github.com/kubernetes/kubernetes/blob/08410cbf06f1c8649157407c660b1c5cb528947b/test/e2e/storage/utils/utils.go#L567
- https://github.com/kubernetes/kubernetes/blob/08410cbf06f1c8649157407c660b1c5cb528947b/test/e2e/storage/utils/utils.go#L71
- https://github.com/kubernetes/kubernetes/blob/08410cbf06f1c8649157407c660b1c5cb528947b/test/e2e/storage/utils/utils.go#L66
- https://github.com/kubernetes/kubernetes/blob/08410cbf06f1c8649157407c660b1c5cb528947b/test/e2e/framework/util.go#L1109
- -> kubectl still needed

@okartau: I'm not sure why these new functions were necessary. Can you use the existing ExecWithOptions/ExecCommandInContainerWithFullOutput from https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/exec_util.go?

Comment (cofyc) [reply test/e2e/framework/pods.go]: yes, this PR is still needed.

`(*Framework).ExecWithOptions/ExecCommandInContainerWithFullOutput` are used to exec command in an existing pod. However, `HostExec` is used to run commands on the host (by using `nsenter` to entering into host namespace). They are for different purposes.

File: test/e2e/storage/utils/utils.go
Comment (cofyc) [inline test/e2e/storage/utils/utils.go]: I think we can use `f.ExecCommandInContainerWithFullOutput` directly.

Comment (okartau) [reply test/e2e/storage/utils/utils.go]: indeed, seems it can, thanks for pointing that out. I changed to that call and patch set became smaller as two added-cloned functions are not needed any more.
Rebased also to HEAD of master and force-pushed.

Comment (pohly) [reply test/e2e/storage/utils/utils.go]: @cofyc: is it okay now?

File: test/e2e/storage/testsuites/volumes.go
Comment (pohly) [inline test/e2e/storage/testsuites/volumes.go]: Looks like we still have a redundant parameter here: `f.ClientSet` can be removed.

**Review by pohly — CHANGES_REQUESTED** [review]

File: test/e2e/storage/utils/utils.go
Comment (cofyc) [reply test/e2e/storage/utils/utils.go]: yes, looks good to me (except one nit you pointed [here](https://github.com/kubernetes/kubernetes/pull/84042#discussion_r344926110))

File: test/e2e/storage/testsuites/volumes.go
Comment (okartau) [reply test/e2e/storage/testsuites/volumes.go]: removed clientset parameter, rebased and force-pushed.

**Comment by pohly** [comment]
This is now entirely in code related to storage testing:
```
$ git diff 5dc87d2919818bbfd5c68eb1aed766b849a9062c..HEAD | diffstat
 common/volumes.go                   |    6 +-
 framework/volume/fixtures.go        |   24 +++++-----
 storage/flexvolume.go               |    8 +--
 storage/persistent_volumes-local.go |   40 ++++++++---------
 storage/testsuites/ephemeral.go     |    8 +--
 storage/testsuites/multivolume.go   |   24 +++++-----
 storage/testsuites/volume_io.go     |   22 ++++-----
 storage/testsuites/volumes.go       |    4 -
 storage/utils/utils.go              |   82 ++++++++++++++++++------------------
 storage/volumes.go                  |    2 
 upgrades/storage/volume_mode.go     |    6 +-
 11 files changed, 114 insertions(+), 112 deletions(-)
```

@msau42, perhaps you can approve?

**Comment by pohly** [comment]
/assign @msau42 
for approval

**Pull request merged**
